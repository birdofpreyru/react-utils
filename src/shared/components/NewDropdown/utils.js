/**
 * Auxiliary functions for <Dropdown> component.
 */

import { findIndex, findLastIndex } from 'lodash';

/**
 * Returns option name and value as two elements of array.
 * @param {object|string} option
 * @return {string[]}
 */
export function optionNameValue(option) {
  return typeof option === 'string'
    ? [option, option] : [option.name, option.value];
}

/**
 * Returns option value.
 * @param {object|string} option Option object or string.
 * @return {string}
 */
export function optionValue(option) {
  return typeof option === 'string' ? option : option.value;
}

/**
 * Finds the option by value.
 * @param {string} value Option value.
 * @param {array} options Options array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Index of the matching option, or -1.
 */
function findOptionIndex(value, options, filter) {
  return options.findIndex(
    (item) => optionValue(item) === value && (!filter || filter(item)),
  );
}

/**
 * Finds the next option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Next option index, or -1.
 */
export function findNextOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter);
  if (index >= 0 && ++index < options.length) {
    return filter ? findIndex(options, filter, index) : index;
  }
  return -1;
}

/**
 * Finds the previous option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Previous option index, or -1.
 */
export function findPrevOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter) - 1;
  if (index >= 0 && filter) index = findLastIndex(options, filter, index);
  return index >= 0 ? index : -1;
}
