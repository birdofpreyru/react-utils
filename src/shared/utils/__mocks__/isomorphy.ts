/**
 * Mockup of the isomorphy module.
 */
/* global jest */

import * as IsomorphyM from '../isomorphy';

const MOCK_BUILD_TIMESTAMP = 'Wed, 29 Nov 2017 07:40:00 GMT';

const mock = jest.requireActual<typeof IsomorphyM>('../isomorphy');

mock.buildTimestamp = () => MOCK_BUILD_TIMESTAMP;

module.exports = mock;
