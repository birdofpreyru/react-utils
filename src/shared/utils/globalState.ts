import type { Request } from 'express';

import { type SsrContext, withGlobalStateType } from '@dr.pogodin/react-global-state';

/** Mapping "chunkName" > array of asset paths. */
export type ChunkGroupsT = Record<string, string[]>;

// The type of data object injected by server into generated markup.
export type InjT = {
  CHUNK_GROUPS?: ChunkGroupsT;
  CONFIG?: Record<string, unknown>;
  ISTATE?: unknown;
};

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    REACT_UTILS_INJECTION?: InjT;
  }
}

// TODO: Not 100% sure now, whether it indeed can be replaced by type,
// or do we really need it to be an interface. Keeping the interface for now.
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export interface SsrContextT<StateT> extends SsrContext<StateT> {
  chunkGroups: ChunkGroupsT;
  chunks: string[];

  /** If set at the end of SSR, the rendered will trigger
   *  server-side redirect to this URL (and use the status
   *  code). */
  redirectTo?: string;

  req: Request;
  status: number;
}

const {
  getSsrContext,
} = withGlobalStateType<unknown, SsrContextT<unknown>>();

export {
  getSsrContext,
};
