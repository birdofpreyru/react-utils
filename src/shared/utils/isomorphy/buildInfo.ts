// Encapsulates access to "Build Info" data.

// BEWARE: This should match the type of build info object generated by
// Webpack build (see "/config/webpack/app-base.js"), and currently this
// match is not checked automatically.
export type BuildInfoT = {
  key: string;
  publicPath: string;
  timestamp: string;
  useServiceWorker: boolean;
};

// Depending on the build mode & environment, BUILD_INFO is either a global
// variable defined at the app launch, or it is replaced by the actual value
// by the Webpack build.
declare const BUILD_INFO: BuildInfoT | undefined;

let buildInfo: BuildInfoT | undefined;

// On the client side "BUILD_INFO" should be injected by Webpack. Note, however,
// that in test environment we may need situations were environment is mocked as
// client-side, although no proper Webpack compilation is executed, thus no info
// injected; because of this we don't do a hard environment check here.
if (typeof BUILD_INFO !== 'undefined') buildInfo = BUILD_INFO;

/**
 * In scenarious where "BUILD_INFO" is not injected by Webpack (server-side,
 * tests, etc.) we expect the host codebase to explicitly set it before it is
 * ever requested. As a precaution, this function throws if build info has been
 * set already, unless `force` flag is explicitly set.
 * @param info
 * @param force
 */
export function setBuildInfo(info?: BuildInfoT, force = false): void {
  if (buildInfo !== undefined && !force) {
    throw Error('"Build Info" is already initialized');
  }
  buildInfo = info;
}

/**
 * Returns "Build Info" object; throws if it has not been initialized yet.
 * @returns
 */
export function getBuildInfo(): BuildInfoT {
  if (buildInfo === undefined) {
    throw Error('"Build Info" has not been initialized yet');
  }
  return buildInfo;
}
