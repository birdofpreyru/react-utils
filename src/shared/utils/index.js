import themed, {
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as url from './url';
import * as webpack from './webpack';

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

let JU = null; // eslint-disable-line import/no-mutable-exports
if (isomorphy.IS_SERVER_SIDE
  && (process.env.NODE_CONFIG_ENV || process.env.NODE_ENV) !== 'production'
) {
  /* eslint-disable global-require */
  const path = webpack.requireWeak('path');
  JU = webpack.requireWeak(path.resolve(__dirname, './jest'));
  /* eslint-enable global-require */
}

/**
 * Creates a new async barrier: a Promise instance with its resolve method
 * attached as .resolve field.
 * @return {Promise<>}
 */
function newBarrier() {
  let resolve;
  const promise = new Promise((res) => { resolve = res; });
  promise.resolve = resolve;
  return promise;
}

/**
 * Attempts to execute provided async "action" up to "maxRetries" time with
 * the given "interval". If any of attempts is successful it returns its result,
 * otherwise if all attempts fail it throws the error of last attempt.
 * @param {function} action
 * @param {number} [maxRetries=5] Optional. Maximum number of retries. Defaults
 *  to 5 attempts.
 * @param {number} [interval=1000] Optional. Interval between retries [ms].
 *  Defaults to 1 second.
 * @return {Promise} Resolves to the result of successful operation, or
 *  rejects with the error from the latst failed attempt.
 */
async function withRetries(action, maxRetries = 5, interval = 1000) {
  /* eslint-disable no-await-in-loop */
  for (let n = 1; ; ++n) {
    try {
      return await action();
    } catch (error) {
      if (n < maxRetries) await time.timer(interval);
      else throw error;
    }
  }
  /* eslint-enable no-await-in-loop */
}

export {
  config,
  isomorphy,
  JU,
  newBarrier,
  themed,
  ThemeProvider,
  time,
  url,
  webpack,
  withRetries,
};
