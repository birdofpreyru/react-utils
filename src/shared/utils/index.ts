import themedImpl, {
  COMPOSE,
  PRIORITY,
  type Theme,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

export {
  type Listener,
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

export {
  type Theme,
  config,
  isomorphy,
  themed,
  ThemeProvider,
  time,
  webpack,
};
