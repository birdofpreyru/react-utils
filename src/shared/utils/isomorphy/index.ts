import { getBuildInfo } from './buildInfo';
import { IS_CLIENT_SIDE, IS_SERVER_SIDE } from './environment-check';

/**
 * @ignore
 * @return {string} Code mode: "development" or "production".
 */
function getMode() {
  return process.env.NODE_ENV;
}

/**
 * Returns `true` if development version of the code is running;
 * `false` otherwise.
 */
export function isDevBuild(): boolean {
  return getMode() === 'development';
}

/**
 * Returns `true` if production build of the code is running;
 * `false` otherwise.
 */
export function isProdBuild(): boolean {
  return getMode() === 'production';
}

/**
 * Returns build timestamp of the front-end JS bundle.
 * @return ISO date/time string.
 */
export function buildTimestamp(): string {
  return getBuildInfo().timestamp;
}

export { IS_CLIENT_SIDE, IS_SERVER_SIDE, getBuildInfo };
