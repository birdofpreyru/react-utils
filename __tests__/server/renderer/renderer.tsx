import type { Request, RequestHandler } from 'express';
import { cloneDeep, noop } from 'lodash';
import factory, { SCRIPT_LOCATIONS, isBrotliAcceptable, type ConfigT } from 'server/renderer';
import fs from 'fs';

import serializeJs from 'serialize-javascript';

import {
  type ForceT,
  getGlobalState,
  useGlobalState,
} from '@dr.pogodin/react-global-state';

import { Helmet } from '@dr.pogodin/react-helmet';

import type { Configuration } from 'webpack';

import { getSsrContext } from 'utils/globalState';

import { getBuildInfo, setBuildInfo, type BuildInfoT } from 'utils/isomorphy/buildInfo';

import {
  mockHttpRequest,
  mockHttpResponse,
  mockWebpackConfig,
} from './__assets__/common';

// eslint-disable-next-line @typescript-eslint/no-namespace
declare namespace global {
  let mockFailsForgeRandomGetBytesMethod: boolean | undefined;
}

global.mockFailsForgeRandomGetBytesMethod = false;

jest.mock('node-forge');

const TEST_CONTEXT = `${__dirname}/__assets__`;

const TEST_INITIAL_STATE = {
  stateKey1: 'State Value #1',
  stateKey2: 'State Value #2',
};

const testBuildInfo = JSON.parse(
  fs.readFileSync(`${TEST_CONTEXT}/.build-info`, 'utf8'),
) as BuildInfoT;

afterEach(() => {
  // TODO: It will be better to completely re-load tested modules, and thus not
  // having to reset this manually.
  setBuildInfo(undefined, true);

  global.mockFailsForgeRandomGetBytesMethod = false;
});

// TODO: Can we update this test to generate the webpack stats and attach
// them to locals by the actual run of webpack-dev-middleware, thus allowing
// to catch any issues caused by breaking changes in the middleware?
/**
 * Performs tests with the specified configuration and options.
 * @param {Object} webpackConfig
 * @param {Object} options
 * @return {Promise}
 */
async function coreTest(webpackConfig: Configuration, options = {}) {
  expect(getBuildInfo).toThrowErrorMatchingSnapshot();

  let renderer: RequestHandler | undefined;
  expect(() => {
    renderer = factory(webpackConfig, {
      ...options,
      logger: {
        debug: noop,
        error: noop,
        info: noop,
        log: noop,
        warn: noop,
      },
    });
  }).not.toThrow();
  expect(renderer).toBeInstanceOf(Function);
  expect(getBuildInfo()).toStrictEqual(testBuildInfo);

  try {
    const { render, res } = mockHttpResponse();
    if (renderer) {
      await renderer(
        (mockHttpRequest() as unknown) as Request,
        res,
        (error) => {
          expect(error).toMatchSnapshot();
        },
      );
    }
    expect(render.markup).toMatchSnapshot();
    expect(render.status).toMatchSnapshot();
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
}

test(
  'Base rendering of HTML template',
  async () => {
    await coreTest(mockWebpackConfig(), {});
  },
);

test(
  '"favicon" option',
  async () => {
    await coreTest(mockWebpackConfig(), { favicon: '/path/to/favicon.ico' });
  },
);

test(
  'Config overriding for injection',
  async () => {
    await coreTest(mockWebpackConfig(), {
      beforeRender: (req: Request, sanitizedConfig: ConfigT) => {
        expect(sanitizedConfig).toBeInstanceOf(Object);
        expect(sanitizedConfig).not.toHaveProperty('SECRET');
        return {
          configToInject: {
            ...sanitizedConfig,
            additionalKey: 'Additional Value',
          },
        };
      },
    });
  },
);

test('JS constructs in global state', async () => {
  const Component = () => {
    // TODO: Could use State type, but not a big deal for now.
    useGlobalState<ForceT, string>('test.1', 'defaultValue');
    useGlobalState<ForceT, Set<number>>('set', new Set([1, 2]));
    const state = getGlobalState().get();
    return <div>{serializeJs(state)}</div>;
  };
  await coreTest(mockWebpackConfig(), {
    Application: Component,
    maxSsrRounds: 3,
  });
});

test(
  'Helmet integration works',
  async () => coreTest(mockWebpackConfig(), {
    Application: () => (
      <div>
        <p>
          Hello World!
        </p>
        <Helmet>
          <title>
            Test Page Title
          </title>
          <meta content="Test Page Description" property="description" />
          <link href="https://github.com/birdofpreyru/react-utils" rel="canonical" />
        </Helmet>
        <p>
          Goodbye World!
        </p>
      </div>
    ),
    maxSsrRounds: 3,
    ssrTimeout: Number.MAX_VALUE,
  }),
);

test(
  'Injection of additional JS scripts',
  async () => coreTest(mockWebpackConfig(), {
    beforeRender: () => ({
      extraScripts: [
        '<script>Dummy JS Sript</script>',
        '<script>Another Dummy JS Script</script>',
        {
          code: '<script>Yet another Dummy JS Script</script>',
          location: SCRIPT_LOCATIONS.DEFAULT,
        },
        {
          code: '<script>1-st script after opening <head></script>',
          location: SCRIPT_LOCATIONS.HEAD_OPEN,
        },
        {
          code: '<script>1-st script after opening <body></script>',
          location: SCRIPT_LOCATIONS.BODY_OPEN,
        },
        {
          code: '<script>2-nd script after opening <body></script>',
          location: SCRIPT_LOCATIONS.BODY_OPEN,
        },
        {
          code: '<script>2-nd script after opening <head></script>',
          location: SCRIPT_LOCATIONS.HEAD_OPEN,
        },
      ],
    }),
  }),
);

test(
  'Server-side rendering (SSR); injection of CSS chunks & Redux state',
  async () => coreTest(mockWebpackConfig(), {
    Application: () => {
      const context = getSsrContext()!;
      context.chunks.push('test-chunk-a');
      context.chunks.push('test-chunk-b');
      return <div>Hello Wold!</div>;
    },
    beforeRender: () => ({
      initialState: cloneDeep(TEST_INITIAL_STATE),
    }),
    maxSsrRounds: 3,
    ssrTimeout: Number.MAX_VALUE,
  }),
);

test(
  'Setting of response HTTP status the server-side rendering',
  async () => coreTest(mockWebpackConfig(), {
    Application: () => {
      const context = getSsrContext()!;
      context.status = 404;
      return <div>404 Error Test</div>;
    },
    maxSsrRounds: 3,
    ssrTimeout: Number.MAX_VALUE,
  }),
);

test('Throws in case of forge.random.getBytes(..) failure', async () => {
  global.mockFailsForgeRandomGetBytesMethod = true;
  await coreTest(mockWebpackConfig(), {});
});

it('correctly tests if brotli-encoded responses are acceptable', () => {
  const res = (header: string) => isBrotliAcceptable({
    // TODO: It is fine to use for sure... should we declare it outside the test?
    // eslint-disable-next-line jest/no-conditional-in-test
    get: (name: string) => (name.toLowerCase() === 'accept-encoding'
      ? header : undefined),
  } as Request);

  expect(res('')).toBe(false);
  expect(res('*')).toBe(true);
  expect(res('br')).toBe(true);
  expect(res('gzip')).toBe(false);
  expect(res('gzip,br')).toBe(true);
  expect(res('gzip, br')).toBe(true);
  expect(res('gzip,br,deflate')).toBe(true);
  expect(res('gzip, br ,deflate')).toBe(true);
  expect(res('gzip,br;q=0.5,deflate')).toBe(true);
  expect(res('gzip,br;q=0.0,deflate')).toBe(false);
  expect(res('gzip,br;q=0,deflate')).toBe(false);
  expect(res('gzip, br;q=0 ,deflate')).toBe(false);
  expect(res('*;q=0')).toBe(false);
  expect(res('br;q=0,*;q=0.5')).toBe(true);
});
