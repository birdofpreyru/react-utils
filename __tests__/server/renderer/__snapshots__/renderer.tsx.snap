// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"favicon" option 1`] = `""Build Info" has not been initialized yet"`;

exports[`"favicon" option 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            <link rel="shortcut icon" href="/favicon.ico">
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWKKnXdbyS28Kq8fENTebXTg==">
          </head>
          <body>
            
            <div id="react-view"></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`"favicon" option 3`] = `200`;

exports[`Base rendering of HTML template 1`] = `""Build Info" has not been initialized yet"`;

exports[`Base rendering of HTML template 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWKKnXdbyS28Kq8fENTebXTg==">
          </head>
          <body>
            
            <div id="react-view"></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`Base rendering of HTML template 3`] = `200`;

exports[`Config overriding for injection 1`] = `""Build Info" has not been initialized yet"`;

exports[`Config overriding for injection 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqZtQgFpXlCvPujKA0tS2UkkPMIbM7b/pz0EW+36BdmLrgRMN8VxVAW5sv2JW+cO6AwYVgAUwRGjir5yU8XAnpA">
          </head>
          <body>
            
            <div id="react-view"></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`Config overriding for injection 3`] = `200`;

exports[`Helmet integration works 1`] = `""Build Info" has not been initialized yet"`;

exports[`Helmet integration works 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Test Page Title</title>
            <meta data-rh="true" content="Test Page Description" property="description"/>
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWKKnXdbyS28Kq8fENTebXTg==">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p><p>Goodbye World!</p></div></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`Helmet integration works 3`] = `200`;

exports[`Injection of additional JS scripts 1`] = `""Build Info" has not been initialized yet"`;

exports[`Injection of additional JS scripts 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            <script>1-st script after opening <head></script><script>2-nd script after opening <head></script>
            
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWKKnXdbyS28Kq8fENTebXTg==">
          </head>
          <body>
            <script>1-st script after opening <body></script><script>2-nd script after opening <body></script>
            <div id="react-view"></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            <script>Dummy JS Sript</script><script>Another Dummy JS Script</script><script>Yet another Dummy JS Script</script>
          </body>
        </html>"
`;

exports[`Injection of additional JS scripts 3`] = `200`;

exports[`JS constructs in global state 1`] = `""Build Info" has not been initialized yet"`;

exports[`JS constructs in global state 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true"></title>
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWUL3h5uSa0xzfT5XBfc+dkJg03js0tc5Fdq1/L/xrtEoxXCKrg6lCyYKZwCTky5ib0caXGM8E73+cVSWoPghSnS4NUPLasB0oQjzb4nzzrnCbhluCzJjpPO8adT213jGzJprOm1VjBtqksqp0zqbPGg==">
          </head>
          <body>
            
            <div id="react-view"><div>{&quot;test&quot;:Array.prototype.slice.call({&quot;1&quot;:&quot;defaultValue&quot;,&quot;length&quot;:2}),&quot;set&quot;:new Set([1,2])}</div></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`JS constructs in global state 3`] = `200`;

exports[`Server-side rendering (SSR); injection of CSS chunks & Redux state 1`] = `""Build Info" has not been initialized yet"`;

exports[`Server-side rendering (SSR); injection of CSS chunks & Redux state 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true"></title>
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet"><link href="/test/public/path/test-chunk-a-1511941200000.css" rel="stylesheet"><link href="/test/public/path/test-chunk-b-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWhVyNrVmy/LG117exMdq7DaCVoYm+R3CZ699xVEmfrItuFaz3fVHwBKqFZCxZ+E96E/P6Ag2ZixSdZfUCA9pAZVJZI0yOj8D3aPaMgRX3PjY=">
          </head>
          <body>
            
            <div id="react-view"><div>Hello Wold!</div></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script><script src="/test/public/path/test-chunk-b-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`Server-side rendering (SSR); injection of CSS chunks & Redux state 3`] = `200`;

exports[`Setting of response HTTP status the server-side rendering 1`] = `""Build Info" has not been initialized yet"`;

exports[`Setting of response HTTP status the server-side rendering 2`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true"></title>
            
            <meta name="theme-color" content="#FFFFFF">
            
            <link href="/test/public/path/main-1511941200000.css" rel="stylesheet">
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBpmFzyzzR1PBUaiLsgpu2NHD6qIeBy+8m9Uwh2qPhPl5dqraHvdasjvg4ljMsqyIAJsUcHuR3mSbcwfxAlIGLmWpQJ3zsX7zkwbR5B44BoZQNVCKVnjAI5SVP4sA20EKLosR2pTrPD/VgFYw7v8C/11sJ/+t3Xfv2bKv3Jw5/dX3Iah0MfS/YXrbxQ3dY9Z2wCNOUY7JC+u76UM1C0X8pFdBMfQQMYO1+KECIxoUo8TqNfLBCrAlYYsQVZIc3TRlWKKnXdbyS28Kq8fENTebXTg==">
          </head>
          <body>
            
            <div id="react-view"><div>404 Error Test</div></div>
            <script src="/test/public/path/main-1511941200000.js" type="application/javascript"></script>
            
          </body>
        </html>"
`;

exports[`Setting of response HTTP status the server-side rendering 3`] = `404`;

exports[`Throws in case of forge.random.getBytes(..) failure 1`] = `""Build Info" has not been initialized yet"`;

exports[`Throws in case of forge.random.getBytes(..) failure 2`] = `[Error: Emulated Failure of forge.random.getBytes(..)]`;

exports[`Throws in case of forge.random.getBytes(..) failure 3`] = `""`;

exports[`Throws in case of forge.random.getBytes(..) failure 4`] = `200`;
